File: fsharp/change/Change.fs
module Change

let findFewestCoins coins target = 
    let valid result = result |> List.sum |> (=) target
    let rec test sum nums result maxlen = 
        match nums with
        | _ when List.length result = maxlen -> result
        | [] -> result
        | h::t -> 
            if sum >= h then
                let subresult = test (sum - h) nums (result @ [h]) maxlen
                if not <| valid subresult then
                    test sum t result maxlen
                else
                    let otherresult = test sum t result (List.length subresult - 1)
                    if otherresult |> valid then otherresult else subresult
            else
                test sum t result maxlen
    let sorted = coins |> List.sortDescending
    let result = test target sorted [] target |> List.sort
    if result |> valid then Some result else None

File: csharp/change/Change.cs
using System;
using System.Collections.Generic;
using System.Linq;

public static class Change
{
    public static int[] FindFewestCoins(int[] coins, int target)
    {
        IEnumerable<int> test(int sum, IEnumerable<int> remaining, IEnumerable<int> result, int maxLength)
        {
            if(result.Count() == maxLength || !remaining.Any()) 
                return result;
            
            var head = remaining.First();
            var tail = remaining.Skip(1);
            if(sum < head)
                return test(sum, tail, result, maxLength);
            var subresult = test(sum - head, remaining, result.Append(head), maxLength);
            if (subresult.Sum() != target)
                return test(sum, tail, result, maxLength);
            var otherresult = test(sum, tail, result, subresult.Count() - 1);
            return otherresult.Sum() == target ? otherresult : subresult;
        }

        var ordered = coins.OrderByDescending(o => o);
        var finalResult = test(target, ordered, Enumerable.Empty<int>(), target).OrderBy(o => o);
        if(finalResult.Sum() != target)
            throw new ArgumentException();
        return finalResult.ToArray();
    }
}